src/aggregator/aggregator/leader_flush_mgr.go

```diff
@@ -181,20 +190,69 @@ func (mgr *leaderFlushManager) Prepare(buckets []*flushBucket) (flushTask, time.
 		}
 	}
 
-	durationSinceLastPersist := time.Duration(nowNanos - mgr.lastPersistAtNanos)
-	if mgr.flushedSincePersist && durationSinceLastPersist >= mgr.flushTimesPersistEvery {
-		mgr.lastPersistAtNanos = nowNanos
-		mgr.flushedSincePersist = false
-		flushTimes := mgr.prepareFlushTimesWithLock(buckets, allShards)
-		mgr.flushTimesManager.StoreAsync(flushTimes)
-	}
-
 	if !shouldFlush {
+		// NB: if enough time has passed since last flush time persist, run the
+		// persist by itself.
+		if shouldUpdateFlushTimes {
+			mgr.flushTask.runNextOnCompleteTask = false
+			return mgr.flushTask.onCompleteTask, waitFor
+		}
+
 		return nil, waitFor
 	}
+
 	return mgr.flushTask, waitFor
 }
 
+type runFunctionTask struct {
+	fn func()
+}
+
+func (t *runFunctionTask) Run() {
+	t.fn()
+}
+
+func (mgr *leaderFlushManager) newUpdateFlushTimesTask(
+	buckets []*flushBucket,
+) flushTask {
+	fn := func() {
+		mgr.RLock()
+		if !mgr.flushedSincePersist {
+			mgr.RUnlock()
+			return
+		}
+
+		shards, err := mgr.placementManager.Shards()
+		if err != nil {
+			mgr.RUnlock()
+			mgr.metrics.getShardsError.Inc(1)
+			mgr.logger.Error("unable to determine shards owned by this instance", zap.Error(err))
+			return
+		}
+
+		mgr.RUnlock()
+
+		allShards := shards.All()
+
+		mgr.Lock()
+		defer mgr.Unlock()
+		if !mgr.flushedSincePersist {
+			return
+		}
+
+		mgr.flushedSincePersist = false
+		flushTimes := mgr.prepareFlushTimesWithLock(buckets, allShards)
+		if err := mgr.flushTimesManager.StoreAsync(flushTimes); err != nil {
```

Comment:
store async still seems dangerous when we want to ensure the flush times are stored before resigning.

Reply:
Ah yeah, that's a good call